linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala

```diff
@@ -0,0 +1,458 @@
+   */
+  private def computeItableBuckets(allClasses: List[LinkedClass],
+      isInterface: ClassName => Boolean): Seq[List[ClassName]] = {
+
+    val classes = allClasses.filterNot(_.kind.isJSType)
+
+    /* The algorithm separates the type hierarchy into three disjoint subsets:
+     *
+     * - join types: types with multiple parents (direct supertypes) that have
+     *   only single subtyping descendants:
+     *   `join(T) = {x ∈ multis(T) | ∄ y ∈ multis(T) : y <: x}`
+     *   where multis(T) means types with multiple direct supertypes.
+     * - spine types: all ancestors of join types:
+     *   `spine(T) = {x ∈ T | ∃ y ∈ join(T) : x ∈ ancestors(y)}`
+     * - plain types: types that are neither join nor spine types
+     *
+     * The bucket assignment process consists of two parts:
+     *
+     * **1. Assign buckets to spine types**
+     *
+     * Two spine types can share the same bucket only if they do not have any
+     * common join type descendants.
+     *
+     * Visit spine types in reverse topological order because (from leaves to
+     * root) when assigning a spine type to a bucket, the algorithm already has
+     * complete information about the join/spine type descendants of that spine
+     * type.
+     *
+     * Assign a bucket to a spine type if adding it does not violate the bucket
+     * assignment rule, namely: two spine types can share a bucket only if they
+     * do not have any common join type descendants. If no existing bucket
+     * satisfies the rule, create a new bucket.
+     *
+     * **2. Assign buckets to non-spine types (plain and join types)**
+     *
+     * Visit these types in level order (from root to leaves). For each type,
+     * compute the set of buckets already used by its ancestors. Assign the
+     * type to any available bucket not in this set. If no available bucket
+     * exists, create a new one.
+     */
+
+    def getAllInterfaces(clazz: LinkedClass): List[ClassName] =
+      clazz.ancestors.filter(isInterface)
+
+    val buckets = new mutable.ListBuffer[Bucket]()
+
+    def findOrCreateBucketSuchThat(p: Bucket => Boolean): Bucket = {
+      buckets.find(p).getOrElse {
+        val newBucket = new Bucket()
+        buckets += newBucket
+        newBucket
+      }
+    }
+
+    /** All join type descendants of the class */
+    val joinsOf =
+      new mutable.HashMap[ClassName, mutable.HashSet[ClassName]]()
+
+    /** the buckets that have been assigned to any of the ancestors of the class */
+    val usedOf = new mutable.HashMap[ClassName, mutable.HashSet[Bucket]]()
+    val spines = new mutable.HashSet[ClassName]()
+
+    // Phase 1: Assign buckets to spine types
+    for (clazz <- classes.reverseIterator) {
+      val ifaces = getAllInterfaces(clazz)
+
+      if (ifaces.nonEmpty) {
+        val joins = joinsOf.getOrElse(clazz.name.name, new mutable.HashSet())
+
+        if (joins.nonEmpty) {
```

Comment:
```suggestion
        joinsOf.get(clazz.name.name) match {
          case Some(joins) =>
            assert(joins.nonEmpty)
            // ...
          case None =>
```

Feels odd to create a set merely to then check it is empty.